src/main/java/io/confluent/maven/resolver/ResolveKafkaVersionRangeMojo.java

```diff
@@ -0,0 +1,239 @@
+	private boolean includeSnapshots;
+
+	/**
+	 * The name of the property that will be set to the highest matching version.
+	 */
+	@Parameter
+	private String property;
+
+	@Component
+	private RepositorySystem repoSystem;
+
+	@Parameter(defaultValue = "${repositorySystemSession}", readonly = true)
+	private RepositorySystemSession repoSession;
+
+	@Parameter(defaultValue = "${project.remoteProjectRepositories}", readonly = true)
+	protected List<RemoteRepository> remoteRepositories;
+
+	@Parameter(defaultValue = "${project}", readonly = true)
+	private MavenProject project;
+
+	@Override
+	public void execute() throws MojoExecutionException, MojoFailureException {
+		VersionRangeRequest request = new VersionRangeRequest();
+		request.setRepositories(remoteRepositories);
+		DefaultArtifact artifact = new DefaultArtifact(groupId, artifactId, null, version);
+		request.setArtifact(artifact);
+
+		String constraintText = artifact.toString() + ", " + (includeSnapshots ? "including" : "excluding")
+				+ " snapshots";
+		getLog().info("Resolving range for " + constraintText);
+		try {
+			VersionRangeResult CEResult = repoSystem.resolveVersionRange(repoSession, request);
+			VersionRangeResult CCSResult = repoSystem.resolveVersionRange(repoSession, request);
+			Version highestCEVersion = fetchHighestKafkaVersion(CE, CEResult, constraintText);
+			Version highestCCSVersion = fetchHighestKafkaVersion(CCS, CCSResult, constraintText);
+
+			getLog().info("Highest " + CE + " version: " + highestCEVersion);
+			getLog().info("Highest " + CCS + " version: " + highestCCSVersion);
+
+			if (printCE) {
+				System.out.println(highestCEVersion);
+			}
+
+			if (printCCS) {
+				System.out.println(highestCCSVersion);
+			}
+
+			// Set CE property.
+			if (StringUtils.isNotEmpty(CEKAFKAVERSION) && project != null) {
+				getLog().info("Setting " + CEKAFKAVERSION + " property " + CEKAFKAVERSION + "=" + highestCEVersion);
+				project.getProperties().put(CEKAFKAVERSION, highestCEVersion.toString());
+			}
+
+			// Set CCS property.
+			if (StringUtils.isNotEmpty(CCSKAFKAVERSION) && project != null) {
+				getLog().info("Setting " + CCSKAFKAVERSION + " property " + CCSKAFKAVERSION + "=" + highestCCSVersion);
+				project.getProperties().put(CCSKAFKAVERSION, highestCCSVersion.toString());
+			}
+		} catch (VersionRangeResolutionException e) {
+			throw new MojoExecutionException("", e);
+		}
+	}
+
+	private Version fetchHighestKafkaVersion(String kafkaType, VersionRangeResult result, String constraintText)
+			throws MojoExecutionException, MojoFailureException {
+
+		// Workaround for https://issues.apache.org/jira/browse/MNG-3092
+		try {
+			getLog().debug(kafkaType + " version fetched " + result.getVersions());
+			if (!includeSnapshots) {
+				result.setVersions(removeSnapshots(result.getVersions()));
+			}
+			if (kafkaType.equals(CE)) {
+				result.setVersions(includeCE(result.getVersions()));
```

Comment:
i don't think this works. as we discussed this getting all of the versions that match the range, so ce and ccs and then excluding one or the other, but the sorting that is still being done by maven core and that will still be lexicographical. so 6.0.0-100-ce will be considered a lower version than 6.0.0-2-ce.

Reply:
At least when i tested locally, the 6.0.0-10000-ce is prior than 6.0.0-12-ce, I might not sure how to mock those maven api call in unit tests, I will see if I can make some unit tests there. The worst case I can always try some IT tests.